src/network/protocol.rs

```diff
@@ -709,16 +708,10 @@ impl<T: Config> UdpProtocol<T> {
 
     /// Upon receiving a `ChecksumReport`, add it to the checksum history
     fn on_checksum_report(&mut self, body: &ChecksumReport) {
-        if self.last_added_checksum_frame < body.frame {
-            if self.checksum_history.len() > MAX_CHECKSUM_HISTORY_SIZE {
-                // keep the checksums later than last_added_checksum_frame - max_checksum_size
-                self.checksum_history.retain(|&frame, _| {
-                    frame > self.last_added_checksum_frame - MAX_CHECKSUM_HISTORY_SIZE as i32
-                });
-            }
-            self.last_added_checksum_frame = body.frame;
-            self.checksum_history.insert(body.frame, body.checksum);
-        }
+        self.pending_checksums
+            .truncate(MAX_CHECKSUM_HISTORY_SIZE - 1);
+        self.pending_checksums
+            .push_front((body.frame, body.checksum));
```

Comment:
@gschup Can these reports arrive out of order?

Reply:
Anything sent over udp can potentially arrive in the wrong order. network/protocol.rs does nothing to prevent that. You can test that using clumsy and selecting the out-of-order packet option. 

I saw you changed the HashMap to a tuple queue. Are you assuming the checksums will always arrive in order?

Reply:
Converted back to a HashMap